### PR TITLE
assert() honor NDEBUG; replace TEE_ASSERT() with panic_unless()

### PR DESCRIPTION
**Pull request by etiennec-linaro**
The assert() macro should expand to a no-op when NDEBUG is defined.
Current implementation provides a TEE_ASSERT() that aborts execution
(if condition fails) whatever debug configuration is. For clarity,
TEE_ASSERT() is renamed panic_unless() and declared in panic.h.

This patch removes several inclusions on tee_common_unpg.h as it used
to declare TEE_ASSERT() and now TEE_ASSERT() is declared in panic.h.
This change lead to including few string.h here and there.

Signed-off-by: jerome forissier <jerome.forissier@linaro.org>
Signed-off-by: etienne carriere <etienne.carriere@linaro.org>

### CODE REVIEW

File: core/arch/arm/include/kernel/static_ta.h [hunk 1]
```diff
@@ -63,7 +63,7 @@ static inline bool is_static_ta_ctx(struct tee_ta_ctx *ctx)
 
 static inline struct static_ta_ctx *to_static_ta_ctx(struct tee_ta_ctx *ctx)
 {
-	assert(is_static_ta_ctx(ctx));
+	panic_unless(is_static_ta_ctx(ctx));
```
Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/static_ta.h]: Please keep the assert.

File: core/arch/arm/include/kernel/user_ta.h [hunk 2]
```diff
@@ -74,7 +74,7 @@ static inline bool is_user_ta_ctx(struct tee_ta_ctx *ctx)
 
 static inline struct user_ta_ctx *to_user_ta_ctx(struct tee_ta_ctx *ctx)
 {
-	assert(is_user_ta_ctx(ctx));
+	panic_unless(is_user_ta_ctx(ctx));
```
Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/user_ta.h]: Please keep the assert

File: core/arch/arm/kernel/generic_boot.c [hunk 3]
```diff
@@ -88,7 +88,7 @@ __weak void main_init_gic(void)
 #if defined(CFG_WITH_ARM_TRUSTED_FW)
 void init_sec_mon(unsigned long nsec_entry __maybe_unused)
 {
-	assert(nsec_entry == PADDR_INVALID);
+	panic_unless(nsec_entry == PADDR_INVALID);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_boot.c]: Please keep the assert

File: core/arch/arm/kernel/generic_boot.c [hunk 4]
```diff
@@ -97,7 +97,7 @@ __weak void init_sec_mon(unsigned long nsec_entry)
 {
 	struct sm_nsec_ctx *nsec_ctx;
 
-	assert(nsec_entry != PADDR_INVALID);
+	panic_unless(nsec_entry != PADDR_INVALID);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_boot.c]: Please keep the assert

File: core/arch/arm/kernel/generic_boot.c [hunk 5]
```diff
@@ -185,8 +185,8 @@ static void init_runtime(unsigned long pageable_part)
 	uint8_t *hashes;
 	size_t block_size;
 
-	TEE_ASSERT(pageable_size % SMALL_PAGE_SIZE == 0);
-	TEE_ASSERT(hash_size == (size_t)__tmp_hashes_size);
+	panic_unless(pageable_size % SMALL_PAGE_SIZE == 0);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_boot.c]: These two are good candidates for `assert()` instead.

File: core/arch/arm/kernel/generic_boot.c [hunk 6]
```diff
@@ -302,15 +302,15 @@ static void init_runtime(unsigned long pageable_part)
 	 */
 	mm = tee_mm_alloc2(&tee_mm_vcore, tee_mm_vcore.lo,
 			(vaddr_t)(__text_init_start - tee_mm_vcore.lo));
-	TEE_ASSERT(mm);
+	panic_unless(mm);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_boot.c]: What makes this `tee_mm_alloc2()` more important than the others?

File: core/arch/arm/kernel/thread.c [hunk 7]
```diff
@@ -259,9 +259,9 @@ struct thread_core_local *thread_get_core_local(void)
 	 * we otherwise may be rescheduled to a different core in the
 	 * middle of this function.
 	 */
-	assert(thread_get_exceptions() & THREAD_EXCP_IRQ);
+	panic_unless(thread_get_exceptions() & THREAD_EXCP_IRQ);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: Please keep the assert

**Comment by jenswi-linaro** [comment]
Lots of `assert()` has been converted into `panic_unless()`. I don't quite agree in most of the cases (where the condition we're asserting is fairly static).

I'd say that in general we're fairly consistent by using `TEE_ASSERT()` where the condition should always be checked and `assert()` where it more something for debug builds.

File: core/arch/arm/kernel/generic_boot.c (hunk 6 above)
Comment (etiennec-linaro) [inline core/arch/arm/kernel/generic_boot.c]: As all other tee_mm allocations are use to access data, core will crash for shure, but this alloc is never used, hence it should be tested. But i agree the resulting C code sequence is somewhat strange.

**Comment by etiennec-linaro** [comment]
I think `panic_unless()` is quite a weird wording.
I Will propose a `panic_if()` instead that is more  explicit but require  switching all related  tests to their negation. Means a careful review.

Thanks for pointing out sequences that should assert or panic. Long task.

I Will also  propose a `panic_msg()` to panic with a specific message as jerome sugested.
I tried to bring a printf like  `panic_msg()`  but it's hard when trying to rely on simple EMSG macros.
For sake of  simplicity i plan only null  terminating string messages.

File: core/arch/arm/kernel/thread.c [hunk 8]
```diff
@@ -167,24 +167,24 @@ void thread_check_canaries(void)
 	size_t n;
```
Comment (etiennec-linaro) [inline core/arch/arm/kernel/thread.c]: For all these canaries i choose the panic way since canaries are weak but ultimated detection of a major failure.

Please tell if ones do not agree.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: It make more sense to have `panic_unless()` for the canaries than the old `assert()`.

**Comment by jenswi-linaro** [comment]
I'd prefer `panic_unless()` over `panic_if()` just because of the negated expressions.

**Comment by etiennec-linaro** [comment]
panic_unless() lead to less changes for shure.
But for source code to be easily readable, i really think a positive test is more explicit than a negative test.  panic_if() or panic_on() is more friendly than a panic_if_not().

**Comment by etiennec-linaro** [comment]
i have updated with a `panic_if()` version. 
* i could revert to a `panic_ifnot()` or 'panic_unless()`.
* are there still places where `assert()` or `panic_xxx()` are misplaced ?

File: mk/config.mk [hunk 9]
```diff
@@ -38,8 +38,10 @@ WARNS ?= 3
 # Define DEBUG=1 to compile with -g option
 # DEBUG=1
 
-# If 1, debug mode of the tee firmware (CPU restart, Core Status)
-CFG_TEE_CORE_DEBUG ?= 0
+# Core debug mode enablement
+# When defined to "y", tee firmware is in debug mode. Core and pager status,
+# panics verbosity, physical/virtual address paranoid verifications.
+CFG_TEE_CORE_DEBUG ?= n
```
Comment (etiennec-linaro) [inline mk/config.mk]: question: `CFG_TEE_CORE_DEBUG` was mis used: tested against ifdef and/or ifeq/ifneq 0.
i wanted to rationalize.
but maybe switching from 0/1 to y/n break some backward compatibility ?

Comment (jenswi-linaro) [inline mk/config.mk]: I wouldn't mind having that changed. But it would be better to take in a separate PR which I assume could be merged before this PR.

**Comment by jenswi-linaro** [comment]
@jforissier, @jbech-linaro how would you like to do with these panics? In my view I'd actually prefer no combined panic macro at all, instead I'd prefer:
```
if (expr)
        panic_msg("Some panic reason if we're not using panic() instead");
```
As it becomes very clear what's going on and it may also discourage the use of panic.

File: mk/config.mk (hunk 9 above)
Comment (etiennec-linaro) [inline mk/config.mk]: ok

**Comment by etiennec-linaro** [comment]
Closing, replaced by [PR#910](https://github.com/OP-TEE/optee_os/pull/910).

Cleaning CFG_TEE_CORE_DEBUG will be done seperatly.
